README.md
# BANS-5GC-in-a-Box

Bandwidth-Allocated Network Slicing on 5G Core

## Introduction

BANS-5GC enables network slicing with specific bandwidth allocation policies on 5G core network by utilizing programmable data plane on transport network.

The 5GC uses [free5GC](https://www.free5gc.org/), an open-source 5G core network.

The programmable data plane application for bandwidth management uses a research from 2019 20th APNOMS[1].

All functions are containerized and deployed on [Kubernetes](https://github.com/kubernetes/kubernetes).

## Supported Components

- Docker v18.06.2
- Kubernetes v1.16.2
- Calico v3.8
- Helm 3.0.0
- free5GC Stage 1 and Stage 2

## Hardware Requirement

- CPU 2 cores
- RAM 8GB
- NIC 2 cards (connecting to eNodeB and the Internet)

## Usage

### Pre-install

```ShellSession
# On Kubernetes node

sudo apt install -y make
```

### Environment Setup

Currently there are no gNodeB and UE for standalone 5GC available in the market yet. Procedure tests are provided for free5GC Stage 2. Skip this section if deploying Stage 2.

S1AP/SCTP protocol for 4GC is still remained between RAN and AMF in free5GC stage 1. Therefore, the following setup is required for field tests.

The setup assumes the eNodeB is directly connected to the server. Please modify eNodeB network configuration to work with your network environment.

#### Without SR-IOV

```ShellSession
# On Kubernetes node

# Configure network environment
sudo ip address add 192.168.3.2/24 dev ${ENODEB_INTF}
sudo sh -c 'echo 1 > /proc/sys/net/ipv4/ip_forward'
```

Configure eNodeB settings
- IP address in subnet 192.168.3.0/24 (excluding ${ENODEB_INTF})
- Gateway/Router 192.168.3.2
- MME/AMF IP 192.168.2.2

NOTE 1: ${ENODEB_INTF} is server's interface connecting to eNodeB.

NOTE 2: For [deploying 5GC only](#deploying-5gc-only), it should be fine to set your desired address and subnet on both eNodeB and Kubernetes node interface connecting to eNodeB. Simply make sure packets can be correctly forwarded between two devices.

NOTE 3: For deploying 5GC [with SDN-based transport](#deploying-5gc-with-sdn-based-transport) and [with BANS](#deploying-5gc-with-bans), ${ENODEB_INTF} should be any available address under subnet 192.168.3.0/24 by default. See [Customizing Configuration](#customizing-configuration) section to customize the subnet in deployment.

NOTE 4: For deploying 5GC [with SDN-based transport](#deploying-5gc-with-sdn-based-transport) and [with BANS](#deploying-5gc-with-bans), it's important to make sure the index of network interface connecting to the Internet is smaller than the one connecting to eNodeB.

#### With SR-IOV

Configure eNodeB settings
- IP address in subnet 192.168.3.0/24 (excluding 192.168.3.2 and 192.168.3.6)
- MME/AMF IP 192.168.3.2

NOTE 1: 192.168.3.2 and 192.168.3.6 are AMF's and UPF's IP address respectively by default. See [Customizing Configuration](#customizing-configuration) section to customize two IP addresses.

### Deploying free5GC Stage 1 Only

```ShellSession
# Deploy
make

# or
make bans-5gcv1
```

### Deploying free5GC Stage 1 with SDN-based Transport

A OvS-based Mininet pod is placed between eNodeB and UPF pod.

Open vSwitches are controlled by [ONOS](https://github.com/opennetworkinglab/onos).

```ShellSession
# Deploy
make bans-5gcv1-ovs
```

### Deploying free5GC Stage 1 with BANS

```ShellSession
# Deploy
make bans-5gcv1-bmv2

# Override default slice configuration
SLICE_CONFIG=/path/to/file make bans-5gcv1-bmv2

# Add new onos slice on existing deployment
SLICE_CONFIG=/path/to/file make onos-bw-slice
```

Example slice configuration is placed at `deploy/slice.json`.

### Deploying free5GC Stage 1 Only with SR-IOV

```ShellSession
# Deploy
# The argument `SRIOV_INTF` is required for server setup and creating SR-IOV resources
SRIOV_INTF=devicename make bans-5gcv1-sriov
```

NOTE 1: The *make* script performs server setup for SR-IOV, such as loading device's kernel module and creating required virtual functions, which is experimental and only be tested with Intel Ethernet adapters. Manually configure your server for SR-IOV devices if *make* fails at target *sriov-server-setup*.

### Deploying free5GC Stage 2 Only

```ShellSession
# Deploy
make bans-5gcv2
```

#### Procedure Test

```ShellSession
# Enter RANSIM pod
export RANSIM_POD=$( kubectl get pod -l app.kubernetes.io/instance=free5gc -l app.kubernetes.io/name=ransim -o jsonpath='{.items[0].metadata.name}' )
kubectl exec -ti $RANSIM_POD bash

# Run procedure test in RANSIM pod

# If deploying with network slicing disabled
cd src/test
go test -v -vet=off -run TestRegistration

# If deploying with network slicing enabled
# Replace <ue-id> with slice ID in configuration .global.slices
cd src/test
go test -v -vet=off -run TestRegistration -ue-id=<ue-id>
```

NOTE 1: For deploying with network slicing enabled, the Registration procedure tests should be executed in the incremental order of *\<ue-id\>*, since NGAP IDs in registration are expected to be incremented strictly.

### Customizing Configuration

```ShellSession
# Customize configuration
BANSVALUES=/path/to/file make [target]
```

Default configuration files for deployment locate in `helm-charts/configs/<target>.yaml`.

See `helm-charts/configs/README.md` for more informations.

### User Interface

Web UI to subscriber database is running on port 30300. Default username and password are *admin* and *1423*.

ONOS GUI is exposed on port 30181. Default username and password are *onos* and *rocks*.

## Troubleshooting

### Unreachable UPF

**Description**

Especially in deploying 5GC [with SDN-based transport](#deploying-5gc-with-sdn-based-transport) and [with BANS](#deploying-5gc-with-bans), ARP cache of UPF pod is incompleted on the Kubernetes node.

**Root Cause**

Calico CNI exploits routing table and iptables infrastructure for communication of workloads, and Kernel sets source IP address of ARP requests for workloads to address of the first available network interface in the order of interface index numbers, except loopback addresses.

The problem takes place when network interface connecting to eNodeB has the smallest index number among all interfaces with available addresses, which means it takes priority over others on the decision of source address of ARP requests. Since UPF pod has a route to direct traffic with the destination of eNodeB to Mininet in such two deployments, replies to ARP requests from UPF cannot be correctly sent back to Kubernetes node.

**Workaround**

Add a dummy address, other than loopback addresses, on the loopback interface to force Kernel set source address of ARP requests to it, since loopback interface has the smallest index number.

## Images

Please see [stevenchiu30801/bans-5gc-images](https://github.com/stevenchiu30801/bans-5gc-images.git).

Note that ONOS image with the bandwidth management application and free5GC image of unreleased version are kept private. Please checkout branch release-1.2 for public release v2.0.2 of free5GC deployment.

## Reference
[1] [P4-Enabled Bandwidth Management](https://ieeexplore.ieee.org/abstract/document/8892909)\
[2] [opencord/automation-tools](https://github.com/opencord/automation-tools)


CHANGELOG.md
# Release-0.1

- Add 5GC deployment only using bare Kubernetes YAML files

# Release-0.2

- Add 5GC deployment with SDN-based transport with bare Kubernetes YAML files

# Release-1.0

- Add 5GC deployment with BANS
- Use Helm 3.0.0 to deploy Kubernetes objects

# Release-1.1

- Experimental server setup script for SR-IOV configuration
- Add 5GC deployment only with SR-IOV
- Add configuration documentation

# Release-1.2

- Rename free5GC Stage 1 target from bans-5gc to bans-5gcv1
- Add free5GC Stage 2
- Remove free5gc-stage-1 git submodule
- Extract Docker images to repository [stevenchiu30801/bans-5gc-images](https://github.com/stevenchiu30801/bans-5gc-images)

# Release-1.3

- Use unreleased version of free5GC for network slicing support
- Support network slicing in AMF, NSSF, SMF, UPF
- Use UPF address in NGAP-PDU Session Setup requests as GTP-U destination in RANSIM tests
- Set UE configuration with network slices in RANSIM tests


helm-charts/free5gc-stage-2/charts/ransim/testscripts/ngsetup_test.go
package test_test

import (
	"fmt"
	"free5gc/lib/ngap"
	"free5gc/lib/ngap/ngapSctp"
	"free5gc/src/test"
	"net"
	"testing"

	"github.com/ishidawataru/sctp"
	"github.com/stretchr/testify/assert"
)

func getNgapIp(amfIP, ranIP string, amfPort, ranPort int) (amfAddr, ranAddr *sctp.SCTPAddr, err error) {
	ips := []net.IPAddr{}
	if ip, err1 := net.ResolveIPAddr("ip", amfIP); err1 != nil {
		err = fmt.Errorf("Error resolving address '%s': %v", amfIP, err1)
		return
	} else {
		ips = append(ips, *ip)
	}
	amfAddr = &sctp.SCTPAddr{
		IPAddrs: ips,
		Port:    amfPort,
	}
	ips = []net.IPAddr{}
	if ip, err1 := net.ResolveIPAddr("ip", ranIP); err1 != nil {
		err = fmt.Errorf("Error resolving address '%s': %v", ranIP, err1)
		return
	} else {
		ips = append(ips, *ip)
	}
	ranAddr = &sctp.SCTPAddr{
		IPAddrs: ips,
		Port:    ranPort,
	}
	return
}

func connectToAmf(amfIP, ranIP string, amfPort, ranPort int) (*sctp.SCTPConn, error) {
	amfAddr, ranAddr, err := getNgapIp(amfIP, ranIP, amfPort, ranPort)
	if err != nil {
		return nil, err
	}
	conn, err := sctp.DialSCTP("sctp", ranAddr, amfAddr)
	if err != nil {
		return nil, err
	}
	info, _ := conn.GetDefaultSentParam()
	info.PPID = ngapSctp.NGAP_PPID
	err = conn.SetDefaultSentParam(info)
	if err != nil {
		return nil, err
	}
	return conn, nil
}

func TestNGSetup(t *testing.T) {
	var n int
	var sendMsg []byte
	var recvMsg = make([]byte, 2048)

	// RAN connect to AMF
	conn, err := connectToAmf("127.0.0.1", "127.0.0.1", 38412, 9487)
	assert.Nil(t, err)

	// send NGSetupRequest Msg
	sendMsg, err = test.GetNGSetupRequest([]byte("\x00\x01\x02"), 24, "free5gc")
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// receive NGSetupResponse Msg
	n, err = conn.Read(recvMsg)
	assert.Nil(t, err)
	_, err = ngap.Decoder(recvMsg[:n])
	assert.Nil(t, err)

	// close Connection
	conn.Close()
}


helm-charts/free5gc-stage-2/charts/ransim/testscripts/registration_test.go
package test_test

import (
	"encoding/hex"
	"gofree5gc/lib/CommonConsumerTestData/UDM/TestGenAuthData"
	"gofree5gc/lib/CommonConsumerTestData/UDR/TestRegistrationProcedure"
	"gofree5gc/lib/MongoDBLibrary"
	"gofree5gc/lib/nas/nasMessage"
	"gofree5gc/lib/nas/nasTestpacket"
	"gofree5gc/lib/nas/nasType"
	"gofree5gc/lib/ngap"
	"gofree5gc/lib/ngap/ngapType"
	"gofree5gc/lib/openapi/models"

	// "gofree5gc/src/ausf/ausf_context"
	"gofree5gc/src/test"

	"flag"
	"fmt"
	"net"
	"os"
	"testing"
	"time"

	"golang.org/x/net/icmp"
	"golang.org/x/net/ipv4"

	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
	"github.com/stretchr/testify/assert"
)

const ranIpAddr string = "{{ .Values.addr }}"

var (
	ueId    int
	icmpCnt int
	icmpDst string
)

var ueData test.UeData

func init() {
	flag.IntVar(&ueId, "ue-id", 0, "UE ID")
	flag.IntVar(&icmpCnt, "icmp-cnt", 5, "ICMP packet count to be sent")
	flag.StringVar(&icmpDst, "icmp-dst", "8.8.8.8", "Destination address of ICMP packets")
	flag.Parse()

	// default ue
	test.Ues = append(test.Ues, test.UeData{
		Supi:        "imsi-2089300007487",
		RanUeNgapId: 1,
		AmfUeNgapId: 1,
		MobileIdentity5GS: nasType.MobileIdentity5GS{
			Len:    12,
			Buffer: []uint8{0x01, 0x02, 0xf8, 0x39, 0xf0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x47, 0x78},
		},
		Sst: 1,
		Sd:  "010203",
	})

	// parse ue configuration
	err := test.ParseConfig("uecfg.yaml")
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	ueData = test.Ues[ueId]
}

func getAuthSubscription() (authSubs models.AuthenticationSubscription) {
	authSubs.PermanentKey = &models.PermanentKey{
		PermanentKeyValue: TestGenAuthData.MilenageTestSet19.K,
	}
	authSubs.Opc = &models.Opc{
		OpcValue: TestGenAuthData.MilenageTestSet19.OPC,
	}
	authSubs.Milenage = &models.Milenage{
		Op: &models.Op{
			OpValue: TestGenAuthData.MilenageTestSet19.OP,
		},
	}
	authSubs.AuthenticationManagementField = "8000"

	authSubs.SequenceNumber = TestGenAuthData.MilenageTestSet19.SQN
	authSubs.AuthenticationMethod = models.AuthMethod__5_G_AKA
	return
}

func getAccessAndMobilitySubscriptionData() (amData models.AccessAndMobilitySubscriptionData) {
	return TestRegistrationProcedure.TestAmDataTable[TestRegistrationProcedure.FREE5GC_CASE]
}

func getSmfSelectionSubscriptionData() (smfSelData models.SmfSelectionSubscriptionData) {
	return TestRegistrationProcedure.TestSmfSelDataTable[TestRegistrationProcedure.FREE5GC_CASE]
}

func getAmPolicyData() (amPolicyData models.AmPolicyData) {
	return TestRegistrationProcedure.TestAmPolicyDataTable[TestRegistrationProcedure.FREE5GC_CASE]
}

func checksum(data []byte) uint16 {
	var (
		sum    uint32
		length int = len(data) - 1
	)

	for i := 0; i < length; i += 2 {
		sum += uint32(data[i]) << 8
		sum += uint32(data[i+1])
	}

	if len(data)%2 == 1 {
		sum += uint32(data[length]) << 8
	}

	sum += sum >> 16

	return ^uint16(sum)
}

func buildGTPHeader(teid uint32, seq uint16) ([]byte, error) {
	const length uint16 = 52
	gtpheader := &layers.GTPv1U{
		Version:             1,
		ProtocolType:        1,
		Reserved:            0,
		ExtensionHeaderFlag: false,
		SequenceNumberFlag:  true,
		NPDUFlag:            false,
		MessageType:         255,
		MessageLength:       length,
		TEID:                teid,
		SequenceNumber:      seq,
	}
	buf := gopacket.NewSerializeBuffer()
	opts := gopacket.SerializeOptions{}
	err := gtpheader.SerializeTo(buf, opts)

	if err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// Registration
func TestRegistration(t *testing.T) {
	var n int
	var sendMsg []byte
	var recvMsg = make([]byte, 2048)

	// set MongoDB
	MongoDBLibrary.SetMongoDB("free5gc", "mongodb://{{ .Values.global.dbServiceDomain }}:27017")

	// RAN connect to AMF
	//// HELM: Remove spaces in substitution to avoid unclosed action error
	//// It seems to only occur when multiple substitution in one line
	conn, err := connectToAmf("{{.Values.amf.ngap.addr}}", "{{.Values.addr}}", 38412, 9487)
	assert.Nil(t, err)

	// send NGSetupRequest Msg
	sendMsg, err = test.GetNGSetupRequest([]byte("\x00\x01\x02"), 24, "free5gc")
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// receive NGSetupResponse Msg
	n, err = conn.Read(recvMsg)
	assert.Nil(t, err)
	_, err = ngap.Decoder(recvMsg[:n])
	assert.Nil(t, err)

	// New UE
	ue := test.NewRanUeContext(ueData.Supi, 1, test.ALG_CIPHERING_128_NEA2, test.ALG_INTEGRITY_128_NIA2)
	// ue := test.NewRanUeContext("imsi-2089300007487", 1, test.ALG_CIPHERING_128_NEA0, test.ALG_INTEGRITY_128_NIA0)
	ue.AmfUeNgapId = ueData.AmfUeNgapId
	ue.RanUeNgapId = ueData.RanUeNgapId
	ue.AuthenticationSubs = getAuthSubscription()
	// insert UE data to MongoDB

	servingPlmnId := "20893"
	test.InsertAuthSubscriptionToMongoDB(ue.Supi, ue.AuthenticationSubs)
	getData := test.GetAuthSubscriptionFromMongoDB(ue.Supi)
	assert.NotNil(t, getData)
	{
		amData := getAccessAndMobilitySubscriptionData()
		test.InsertAccessAndMobilitySubscriptionDataToMongoDB(ue.Supi, amData, servingPlmnId)
		getData := test.GetAccessAndMobilitySubscriptionDataFromMongoDB(ue.Supi, servingPlmnId)
		assert.NotNil(t, getData)
	}
	{
		smfSelData := getSmfSelectionSubscriptionData()
		test.InsertSmfSelectionSubscriptionDataToMongoDB(ue.Supi, smfSelData, servingPlmnId)
		getData := test.GetSmfSelectionSubscriptionDataFromMongoDB(ue.Supi, servingPlmnId)
		assert.NotNil(t, getData)
	}
	{
		amPolicyData := getAmPolicyData()
		test.InsertAmPolicyDataToMongoDB(ue.Supi, amPolicyData)
		getData := test.GetAmPolicyDataFromMongoDB(ue.Supi)
		assert.NotNil(t, getData)
	}

	// send InitialUeMessage(Registration Request)(imsi-2089300007487)
	mobileIdentity5GS := ueData.MobileIdentity5GS
	registrationRequest := nasTestpacket.GetRegistrationRequestWith5GMM(nasMessage.RegistrationType5GSInitialRegistration, mobileIdentity5GS, nil, nil)
	sendMsg, err = test.GetInitialUEMessage(ue.RanUeNgapId, registrationRequest, "")
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// receive NAS Authentication Request Msg
	n, err = conn.Read(recvMsg)
	assert.Nil(t, err)
	ngapMsg, err := ngap.Decoder(recvMsg[:n])
	assert.Nil(t, err)

	// Calculate for RES*
	nasPdu := test.GetNasPdu(ngapMsg.InitiatingMessage.Value.DownlinkNASTransport)
	assert.NotNil(t, nasPdu)
	rand := nasPdu.AuthenticationRequest.GetRANDValue()
	resStat := ue.DeriveRESstarAndSetKey(ue.AuthenticationSubs, rand[:], "5G:mnc093.mcc208.3gppnetwork.org")

	// send NAS Authentication Response
	pdu := nasTestpacket.GetAuthenticationResponse(resStat, "")
	sendMsg, err = test.GetUplinkNASTransport(ue.AmfUeNgapId, ue.RanUeNgapId, pdu)
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// receive NAS Security Mode Command Msg
	n, err = conn.Read(recvMsg)
	assert.Nil(t, err)
	_, err = ngap.Decoder(recvMsg[:n])
	assert.Nil(t, err)

	// send NAS Security Mode Complete Msg
	pdu = nasTestpacket.GetSecurityModeComplete(registrationRequest)
	pdu, err = test.EncodeNasPduWithSecurity(ue, pdu)
	assert.Nil(t, err)
	sendMsg, err = test.GetUplinkNASTransport(ue.AmfUeNgapId, ue.RanUeNgapId, pdu)
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// receive ngap Initial Context Setup Request Msg
	n, err = conn.Read(recvMsg)
	assert.Nil(t, err)
	_, err = ngap.Decoder(recvMsg[:n])
	assert.Nil(t, err)

	// send ngap Initial Context Setup Response Msg
	sendMsg, err = test.GetInitialContextSetupResponse(ue.AmfUeNgapId, ue.RanUeNgapId)
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// send NAS Registration Complete Msg
	pdu = nasTestpacket.GetRegistrationComplete(nil)
	pdu, err = test.EncodeNasPduWithSecurity(ue, pdu)
	assert.Nil(t, err)
	sendMsg, err = test.GetUplinkNASTransport(ue.AmfUeNgapId, ue.RanUeNgapId, pdu)
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	time.Sleep(100 * time.Millisecond)
	// send GetPduSessionEstablishmentRequest Msg

	sNssai := models.Snssai{
		Sst: ueData.Sst,
		Sd:  ueData.Sd,
	}
	pdu = nasTestpacket.GetUlNasTransport_PduSessionEstablishmentRequest(10, nasMessage.ULNASTransportRequestTypeInitialRequest, "internet", &sNssai)
	pdu, err = test.EncodeNasPduWithSecurity(ue, pdu)
	assert.Nil(t, err)
	sendMsg, err = test.GetUplinkNASTransport(ue.AmfUeNgapId, ue.RanUeNgapId, pdu)
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// recieve 12. NGAP-PDU Session Resource Setup Request(DL nas transport((NAS msg-PDU session setup Accept)))
	n, err = conn.Read(recvMsg)
	assert.Nil(t, err)
	ngapPduMsg, err := ngap.Decoder(recvMsg[:n])
	assert.Nil(t, err)

	// parse UPF address from NGAP-PDU Session Resource Setup Request
	var upfAddr string
	// fmt.Printf("%+v\n", ngapPduMsg.InitiatingMessage.Value)
	iEList := ngapPduMsg.InitiatingMessage.Value.PDUSessionResourceSetupRequest.ProtocolIEs.List
	for i := 0; i < len(iEList); i++ {
		// fmt.Printf("%+v\n", iEList[i].Id)
		// fmt.Printf("%+v\n", iEList[i].Value)
		if iEList[i].Value.Present == ngapType.PDUSessionResourceSetupRequestIEsPresentPDUSessionResourceSetupListSUReq {
			sUReqList := iEList[i].Value.PDUSessionResourceSetupListSUReq.List
			for j := 0; j < len(sUReqList); j++ {
				// fmt.Printf("%+v\n", sUReqList[j])
				if sUReqList[j].PDUSessionID.Value == int64(10) {
					// fmt.Printf("%+v\n", sUReqList[j].PDUSessionResourceSetupRequestTransfer)
					// TODO: unmarshal aper.OctetString and get TransportLayerAddress in field ULNGUUPTNLInformation
					upfAddr = net.IP(sUReqList[j].PDUSessionResourceSetupRequestTransfer[9:13]).String()
					break
				}
			}
		}
	}

	// send 14. NGAP-PDU Session Resource Setup Response
	sendMsg, err = test.GetPDUSessionResourceSetupResponse(ue.AmfUeNgapId, ue.RanUeNgapId, ranIpAddr)
	assert.Nil(t, err)
	_, err = conn.Write(sendMsg)
	assert.Nil(t, err)

	// wait 1s
	time.Sleep(1 * time.Second)

	// RAN connect to UPF
	upfConn, err := connectToUpf(ranIpAddr, upfAddr, 2152, 2152)
	assert.Nil(t, err)
	fmt.Printf("UPF %s connected\n", upfAddr)

	// send ICMP packets
	fmt.Printf("Send %d ICMP packet(s) to %s\n", icmpCnt, icmpDst)
	for i := 0; i < icmpCnt; i++ {
		// build GTP header
		gtpHdrBuf, err := buildGTPHeader(1, uint16(i))
		assert.Nil(t, err)

		// build IPv4 header
		ipv4hdr := ipv4.Header{
			Version:  4,
			Len:      20,
			Protocol: 1,
			Flags:    0,
			TotalLen: 48,
			TTL:      64,
			Src:      net.ParseIP("60.60.0.1").To4(),
			Dst:      net.ParseIP(icmpDst).To4(),
			ID:       1,
			Checksum: 0,
		}
		v4HdrBuf, err := ipv4hdr.Marshal()
		assert.Nil(t, err)

		// compute IP checksum
		ipv4hdr.Checksum = int(checksum(v4HdrBuf))
		v4HdrBuf, err = ipv4hdr.Marshal()
		assert.Nil(t, err)

		// build ICMP packet
		icmpData, err := hex.DecodeString("8c870d0000000000101112131415161718191a1b")
		assert.Nil(t, err)

		icmpPacket := icmp.Message{
			Type: ipv4.ICMPTypeEcho, Code: 0,
			Body: &icmp.Echo{
				ID: 0, Seq: i,
				Data: icmpData,
			},
		}
		icmpBuf, err := icmpPacket.Marshal(nil)
		assert.Nil(t, err)

		// concat headers and payload
		gtpPacket := append(gtpHdrBuf, v4HdrBuf...)
		gtpPacket = append(gtpPacket, icmpBuf...)

		_, err = upfConn.Write(gtpPacket)
		assert.Nil(t, err)
		sentTime := time.Now()

		go func() {
			// receive icmp response
			n, err = upfConn.Read(recvMsg)
			assert.Nil(t, err)
			recvTime := time.Now()
			elapsed := recvTime.Sub(sentTime)

			// decode packet
			var gtpu layers.GTPv1U
			var ip4 layers.IPv4
			var icmp4 layers.ICMPv4
			var payload gopacket.Payload
			parser := gopacket.NewDecodingLayerParser(layers.LayerTypeGTPv1U, &gtpu, &ip4, &icmp4, &payload)
			decoded := []gopacket.LayerType{}
			err = parser.DecodeLayers(recvMsg, &decoded)
			var ttl uint8
			for _, layerType := range decoded {
				switch layerType {
				case layers.LayerTypeIPv4:
					ttl = ip4.TTL
				case layers.LayerTypeICMPv4:
					if icmp4.Seq == uint16(i) {
						fmt.Printf("%d bytes from %s: icmp_seq=%d ttl=%d time=%.2f ms\n", len(payload), icmpDst, i, ttl, float64(elapsed)/float64(time.Millisecond))
					}
				}
			}
			if err != nil {
				fmt.Println("  Error encountered:", err)
			}
		}()

		// wait one second interval
		time.Sleep(1 * time.Second)
	}

	// delete test data
	test.DelAuthSubscriptionToMongoDB(ue.Supi)
	test.DelAccessAndMobilitySubscriptionDataFromMongoDB(ue.Supi, servingPlmnId)
	test.DelSmfSelectionSubscriptionDataFromMongoDB(ue.Supi, servingPlmnId)

	// close Connection
	conn.Close()
}


scripts/sriov_config.sh
#!/bin/bash

usage() {
    echo "***Experimental script for SR-IOV configuration***"
    echo "Configure virtual functions on SR-IOV device"
    echo ""
    echo "Usage: ./sriov_config.sh SRIOV-INTF NUM_VF"
    echo "Arguments:"
    echo "    SRIOV_INTF        Interface to be enabled SR-IOV"
    echo "    NUM_VF            Number of VFs to be created"
}

if [[ $# -eq 2 ]]; then
    if [[ $1 == "-h" ]] || [[ $1 == "--help" ]]; then
        usage
        exit 0
    fi

    # Check if input NUM_VF is an integer
    if ! [[ $2 =~ ^[0-9]+$ ]]; then
        usage
        exit 1
    fi
else
    usage
    exit 1
fi

# Check if number of existing VFs meets requested number
EXISTING_VF=$( ip link show $1 | grep -c vf )
if [[ ${EXISTING_VF} -eq $2 ]]; then
    echo "Number of existing VFs is exactly $2"
    exit 0
fi

# Check if IOMMU support for Linux kernel is enable
if ! dmesg | grep 'DMAR: IOMMU enabled' >/dev/null; then
    echo -e "Please enable IOMMU support for Linux kernel\n\
        1. Append 'intel_iommu=on' to the 'GRUM_CMDLINE_LINUX' entry in /etc/default/grub\n\
        2. Update grub with command 'update-grub'\n\
        3. Reboot for IOMMU change to take effect"
    exit 1
fi

# Get the driver used by the device
DRIVER=$( ethtool -i $1 | awk '$1=="driver:" {print $2}' )

if [[ -z "${DRIVER}" ]]; then
    echo "No driver information for device '$1'"
    exit 1
fi

# Check if device's kernel module is loaded
if ! lsmod | awk '{print $1}' | grep ${DRIVER} >/dev/null; then
    # Check if driver is included in Linux kernel
    if ! find /lib/modules/$(uname -r) -type f -name '*.ko' | grep /${DRIVER}.ko >/dev/null; then
        echo "PF driver '${DRIVER}' is not provided in kernel"
        echo "Please install the driver '${DRIVER}'"
        exit 1
    elif ! find /lib/modules/$(uname -r) -type f -name '*.ko' | grep /${DRIVER}vf.ko >/dev/null; then
        echo "VF driver '${DRIVER}vf' is not provided in kernel"
        echo "Please install the driver '${DRIVER}vf'"
        exit 1
    fi

    # Load device's kernel module
    modprobe ${DRIVER}
fi

# Check if the requested number of VFs exceeds the maximum number of supported VFs
TOTAL_VF=$( cat /sys/class/net/$1/device/sriov_totalvfs )
if [[ $2 -gt ${TOTAL_VF} ]]; then
    echo "Number of VFs should not exceed the maximum number of VFs supported by device '$1'"
    echo "Maximum number: ${TOTAL_VF}"
    exit 1
fi

# Create VFs
echo $2 | sudo tee /sys/class/net/$1/device/sriov_numvfs >/dev/null


deploy/mininet/toposcripts/topo.py
#!/usr/bin/python

import socket
from functools import partial

from mininet.net import Mininet
from mininet.link import Intf
from mininet.topo import Topo
from mininet.node import OVSSwitch, RemoteController
from mininet.cli import CLI
from mininet.util import quietRun
from mininet.log import setLogLevel, info, error

class MinimalTopo(Topo):

    def __init__(self):
        Topo.__init__(self)

        info( '*** Adding switch\n' )
        OVSSwitch13 = partial( OVSSwitch, protocols='OpenFlow13' )
        s1 = self.addSwitch( 's1', cls=OVSSwitch13 )

if __name__ == '__main__':
    setLogLevel( 'info' )

    topo = MinimalTopo()
    net = Mininet( topo=topo, controller=None )

    info( '*** Adding hardware interface to switch s1\n')
    s1 = net.get( 's1' )
    _intf_1 = Intf( 'net1', node=s1 )
    _intf_2 = Intf( 'net2', node=s1 )

    info( '*** Turning off checksum offloading\n' )
    print quietRun( 'ethtool -K net1 tx off rx off' )
    print quietRun( 'ethtool -K net2 tx off rx off' )

    info( '*** Adding controllers\n' )
    controllerIp = socket.gethostbyname( 'onos-openflow-service.default.svc.cluster.local' )
    net.addController( 'onos', controller=RemoteController, ip=controllerIp, port=6653 )

    net.start()
    CLI( net )
    net.stop()


helm-charts/mininet/toposcripts/topo-bmv2.py
#!/usr/bin/python

import socket

from mininet.net import Mininet
from mininet.link import Intf
from mininet.topo import Topo
from mininet.node import RemoteController
from mininet.cli import CLI
from mininet.util import quietRun
from mininet.log import setLogLevel, info, error

from bmv2 import ONOSBmv2Switch

PIPECONF_ID = 'org.onosproject.pipelines.basic'

class MinimalTopo(Topo):

    def __init__(self):
        Topo.__init__(self)

        info( '*** Adding switch\n' )
        controllerIp = socket.gethostbyname( '{{ .Values.onosGuiService }}' )
        s1 = self.addSwitch( 's1', cls=ONOSBmv2Switch, controller=controllerIp, pipeconf=PIPECONF_ID )

if __name__ == '__main__':
    setLogLevel( 'info' )

    topo = MinimalTopo()
    net = Mininet( topo=topo, controller=None )

    s1 = net.get( 's1' )
{{- range $i, $junk := until (.Values.numIntfs|int) -}}
{{- $intf := printf "net%d" (add $i 1) }}

    info( '*** Adding hardware interface {{ $intf }} to switch s1\n')
    _intf_1 = Intf( '{{ $intf }}', node=s1 )

    info( '*** Turning off checksum offloading on {{ $intf }}\n' )
    print quietRun( 'ethtool -K {{ $intf }} tx off rx off' )
{{- end }}

    info( '*** Adding controllers\n' )
    net.addController( 'onos', controller=RemoteController )

    net.start()
    CLI( net )
    net.stop()


helm-charts/free5gc-stage-2/charts/ransim/testscripts/parse_config.go
package test

import (
	"gofree5gc/lib/nas/nasType"

	"encoding/hex"
	"io/ioutil"

	"gopkg.in/yaml.v2"
)

type UeData struct {
	Supi              string `yaml:"supi"`
	RanUeNgapId       int64  `yaml:"ranUeNgapId"`
	AmfUeNgapId       int64  `yaml:"amfUeNgapId"`
	Sst               int32  `yaml:"sst"`
	Sd                string `yaml:"sd"`
	MobileIdentity5GS nasType.MobileIdentity5GS
}

var Ues []UeData

func ParseConfig(f string) error {
	content, err := ioutil.ReadFile(f)
	if err != nil {
		return err
	}

	var ueConfig []UeData
	err = yaml.Unmarshal([]byte(content), &ueConfig)
	if err != nil {
		return err
	}

	// calculate MobileIdentity5GS with MSIN in Little Endian
	for idx, ue := range ueConfig {
		char_1 := ue.Supi[len(ue.Supi)-1:]
		char_2 := ue.Supi[len(ue.Supi)-2 : len(ue.Supi)-1]
		decoded, err := hex.DecodeString(char_1 + char_2)
		if err != nil {
			return err
		}

		buf := []uint8{0x01, 0x02, 0xf8, 0x39, 0xf0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x47}
		buf = append(buf, []uint8(decoded)...)
		ueConfig[idx].MobileIdentity5GS = nasType.MobileIdentity5GS{
			Len:    12,
			Buffer: buf,
		}
	}

	Ues = append(Ues, ueConfig...)

	return nil
}
